File: src/main/java/com/github/kindrat/cassandra/client/ui/fx/CellFactory.java
package com.github.kindrat.cassandra.client.ui.fx;

import com.datastax.driver.core.TypeCodec;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.scene.control.cell.TextFieldTableCell;
import javafx.util.Callback;
import javafx.util.StringConverter;
import lombok.experimental.UtilityClass;
import org.apache.commons.lang3.StringUtils;

@UtilityClass
public class CellFactory {
    public static <S, T> Callback<TableColumn<S, T>, TableCell<S, T>> create(TypeCodec<T> codec) {
        return TextFieldTableCell.forTableColumn(new StringConverter<>() {
            @Override
            public String toString(T object) {
                return codec.format(object);
            }

            @Override
            public T fromString(String string) {
                return codec.parse(StringUtils.wrapIfMissing(string, "'"));
            }
        });
    }
}

File: src/main/kotlin/com/github/kindrat/cassandra/client/ui/fx/CellFactory.kt
package com.github.kindrat.cassandra.client.ui.fx

import com.datastax.oss.driver.api.core.type.codec.TypeCodec
import javafx.scene.control.TableCell
import javafx.scene.control.TableColumn
import javafx.scene.control.cell.TextFieldTableCell
import javafx.util.Callback
import javafx.util.StringConverter
import org.apache.commons.lang3.StringUtils

object CellFactory {
    fun <S, T> create(codec: TypeCodec<T>): Callback<TableColumn<S, T>, TableCell<S, T>> {
        return TextFieldTableCell.forTableColumn(object : StringConverter<T>() {
            override fun toString(value: T?): String {
                return codec.format(value)
            }

            override fun fromString(string: String): T? {
                return codec.parse(StringUtils.wrapIfMissing(string, "'"))
            }
        })
    }
}